File: DSA_Level-1/Basic Data Structures/Stacks And Queues/Postfix Evaluation And Conversions/PostfixEvaluationAndConversions_Solution.java
import java.io.*;
import java.util.*;

public class PostfixEvaluationAndConversions_Solution{
  
public static int process(int v1,int v2,char op){
    //console.log(`${v1}${op}${v2}`);
    if(op=='+'){
        return v1+v2;
    }
    else if(op=='-'){
        return v1-v2;
    }
    else if(op=='*'){
        return v1*v2;
    }
    else{
        return v1/v2;
    }

}    

public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    String str = br.readLine();

    // code
    Stack<Integer> valueStack=new Stack<>();
    Stack<String> infixStack=new Stack<>();
    Stack<String> prefixStack=new Stack<>();
    for(int i=0;i<str.length();++i){
        char ch=str.charAt(i);
        if(ch=='+' || ch=='-' || ch=='*' || ch=='/'){
            int val2=valueStack.pop();
            int val1=valueStack.pop();
            int ans=process(val1,val2,ch);
            valueStack.push(ans);

            String valS2=infixStack.pop();
            String valS1=infixStack.pop();
            String ansS="("+valS1+ch+valS2+")";
            infixStack.push(ansS);

            valS2=prefixStack.pop();
            valS1=prefixStack.pop();
            ansS=""+ch+valS1+valS2;
            prefixStack.push(ansS);
        }
        else{
            valueStack.push(ch-'0');
            infixStack.push(ch+"");
            prefixStack.push(ch+"");
        }
    }

    System.out.println(valueStack.peek());
    System.out.println(infixStack.peek());
    System.out.println(prefixStack.peek());
 }
}

File: DSA_Level-1/Basic Data Structures/Stacks And Queues/Postfix Evaluation And Conversions/PostfixEvaluationAndConversions_Solution.js
class Stack{
    constructor(){
        this.arr=[];
    }

    push(ch){
        this.arr.push(ch);
    }

    pop(){
        return this.arr.pop();
    }

    size(){
        return this.arr.length;
    }

    peek(){
        return this.arr[this.arr.length-1];
    }
}

function process(v1,v2,op){
    //console.log(`${v1}${op}${v2}`);
    if(op=="+"){
        return v1+v2;
    }
    else if(op=="-"){
        return v1-v2;
    }
    else if(op=="*"){
        return v1*v2;
    }
    else{
        return v1/v2;
    }

}

function postfixEvaluationAndConversion(str){
    let valueStack=new Stack();
    let infixStack=new Stack();
    let prefixStack=new Stack();
    for(let i=0;i<str.length;++i){
        let ch=str[i];
        if(ch=='+' || ch=='-' || ch=='*' || ch=='/'){
            let val2=valueStack.pop();
            let val1=valueStack.pop();
            let ans=process(val1,val2,ch);
            valueStack.push(ans);

            val2=infixStack.pop();
            val1=infixStack.pop();
            ans=`(${val1}${ch}${val2})`;
            infixStack.push(ans);

            val2=prefixStack.pop();
            val1=prefixStack.pop();
            ans=`${ch}${val1}${val2}`;
            prefixStack.push(ans);
        }
        else{
            valueStack.push(Number.parseInt(ch));
            infixStack.push(ch);
            prefixStack.push(ch);
        }
    }

    console.log(valueStack.peek());
    console.log(infixStack.peek());
    console.log(prefixStack.peek());
}


postfixEvaluationAndConversion("264*8/+3-");